onnxruntime/core/providers/cpu/tensor/upsample_antialias.h

```diff
@@ -0,0 +1,763 @@
+    }
+
+    // taken from https://github.com/python-pillow/Pillow/blob/66add095a50d76c35c7f58643461f2edf78a3f05/src/libImaging/Resample.c#L94
+    //  we need to handle negative values
+    //  it's equivalent to :x = np.clip(x, 0, 255) where x \in [-640, 639]
+    // we will accept a negative x for (&clip8_lookups_table[640])[x] means clip8_lookups_table +640 -x
+    for (int i = 0; i < 1280; ++i) {
+      clip8_lookups_table[i] = static_cast<uint8_t>(std::min(std::max(i - 640, 0), 255));
+    }
+  }
+  virtual ~FilterParamsAntiAlias() = default;
+  virtual float filter(float x) const = 0;
+};
+
+template <typename T>
+struct BilinearParamsAntiAlias : FilterParamsAntiAlias<T> {
+  // taken from
+  // https://github.com/python-pillow/Pillow/blob/6812205f18ca4ef54372e87e1a13ce4a859434df/
+  // src/libImaging/Resample.c#L20-L29
+  float filter(float x) const override {
+    if (x < 0.0f) {
+      x = -x;
+    }
+    if (x < 1.0f) {
+      return 1.0f - x;
+    }
+    return 0.0f;
+  }
+};
+
+template <typename T>
+struct BiCubicParamsAntiAlias : FilterParamsAntiAlias<T> {
+  BiCubicParamsAntiAlias() {
+    this->support_size = 4.0f;
+  }
+
+  // taken from
+  // https://github.com/python-pillow/Pillow/blob/6812205f18ca4ef54372e87e1a13ce4a859434df/
+  // src/libImaging/Resample.c
+  float filter(float x) const override {
+    /* https://en.wikipedia.org/wiki/Bicubic_interpolation#Bicubic_convolution_algorithm
+     */
+    if (x < 0.0f) {
+      x = -x;
+    }
+    if (x < 1.0f) {
+      return ((this->cubic_coeff_a + 2.0f) * x - (this->cubic_coeff_a + 3.0f)) * x * x + 1;
+    }
+    if (x < 2.0f) {
+      return (((x - 5.0f) * x + 8.f) * x - 4.f) * this->cubic_coeff_a;
+    }
+    return 0.0f;
+  }
+};
+
+template <typename T>
+struct TriLinearParamsAntiAlias : FilterParamsAntiAlias<T> {
+  float filter(float x) const override {
+    if (x < 0.0f) {
+      x = -x;
+    }
+    if (x < 1.0f) {
+      return 1.0f - x;
+    }
+    return 0.0f;
+  }
+};
+
+template <typename T>
+struct AccumulateType {
+  using type = int32_t;
+  using Dtype = T;
+};
+
+template <>
+struct AccumulateType<int32_t> {
+  using type = float;
+};
+
+template <>
+struct AccumulateType<float> {
+  using type = float;
+};
+
+template <>
+struct AccumulateType<double> {
+  using type = double;
+};
+
+// The following method supports a 3/4/5-D input in 'Linear mode, cubic mode'
+// that amounts to 'Bilinear,TriLinear, Bicubic/Tricubic' Upsampling/Resizing in the sense that it assumes
+// A N-D tensor has
+// 1. the scale values for the outermost 2 dimensions are 1 or
+// 2. the scale values for the outermost and innermost dimensions are 1
+// This is the common use-case where the 4-D input (batched multi-channel images)
+// is usually of shapes:
+// - [N, C, H, W] and the scales are [1.0, 1.0, height_scale, width_scale]
+// - [N, H, W, C] and the scales are [1.0, height_scale, width_scale, 1.0]
+template <class T>
+void SetupUpsampleFilterAntiAlias(FilterParamsAntiAlias<T>& p,
```

Comment:
Could we potentially refine this in the future to do the setup once in the ctor if the input values are fixed?

Reply:
It could be. I doubt how many gains we could have though. This function/process should be done very fast.